### ISSUE TITLE
Rule proposal: unreachable-nullish-coalescing-operator

### ISSUE THREAD

**Issue by loynoir** [typescript-eslint/typescript-eslint#7005]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+new+plugin+rule%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] My proposal specifically checks TypeScript syntax, or it proposes a check that requires type information to be accurate.
- [X] My proposal is not a "formatting rule"; meaning it does not just enforce how code is formatted (whitespace, brace placement, etc).
- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

`unreachable-nullish-coalescing-operator` lint unreachable condition for `nullish-coalescing-operator`

### Fail Cases

```typescript
type V0 = string | undefined
type V1 = string

// const t: V0 = '42'
const t: V1 = '42'

console.log(
  t ?? function () {
    throw new Error()
  }(),
)
```

### Pass Cases

```typescript
type V0 = string | undefined
type V1 = string

// const t: V0 = '42'
const t: V1 = '42'

console.log(t)
```

**Comment by bradzacher** [comment]
The no-unnecessary-condition rule already does this.

https://typescript-eslint.io/play/#ts=5.0.4&sourceType=module&code=C4TwDgpgBAagDFAvFAzsATgSwHYHMoA+UArtgCYQBmOEZAUKJLAIxKoY6510D0PUAYwD22NFGAAuWAmQByACwAmWXWGjg4qTFZylK1SJRCANhAB0xobgAUdKOKgB+R1EqkBwTCKjWAlFABvO3txAAt0IQB3KGwIaIBRdAj0P2CAXz8AGjpfOiA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK+SpPRRE0aB2iRwYAL4gtQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA

https://typescript-eslint.io/rules/no-unnecessary-condition

**Comment by loynoir** [comment]
@bradzacher 

Would be nice to add keyword `unreachable` and `nullish-coalescing-operator`.

I searched, and did not found it.

**Comment by Josh-Cena** [comment]
I agree that it's not apparent that `no-unnecessary-condition` checks `??` because people tend not to think of the latter as a "condition", but that would need to be tracked in a separate issue.
